lighthouse-core/audits/byte-efficiency/unused-javascript.js

```diff
@@ -65,43 +72,131 @@ class UnusedJavaScript extends ByteEfficiencyAudit {
+        return {
+          key,
+          unused: Math.round(unusedBytes * transferRatio),
+          total: Math.round(total * transferRatio),
+        };
+      });
+
+    Object.assign(item, {
+      sources: topUnusedFilesSizes.map(d => d.key),
+      sourceBytes: topUnusedFilesSizes.map(d => d.total),
+      sourceWastedBytes: topUnusedFilesSizes.map(d => d.unused),
+    });
+  }
+
+  /**
+   * @param {WasteData[]} wasteData
+   * @param {string} url
+   * @param {ReturnType<typeof UnusedJavaScript.determineLengths>} lengths
    * @return {LH.Audit.ByteEfficiencyItem}
    */
-  static mergeWaste(wasteData, networkRecord) {
-    let unusedLength = 0;
-    let contentLength = 0;
+  static mergeWaste(wasteData, url, lengths) {
+    let unused = 0;
+    let content = 0;
+    // TODO: this is right for multiple script tags in an HTML document,
+    // but may be wrong for multiple frames using the same script resource.
     for (const usage of wasteData) {
-      unusedLength += usage.unusedLength;
-      contentLength += usage.contentLength;
+      unused += usage.unusedLength;
+      content += usage.contentLength;
     }
 
-    const totalBytes = ByteEfficiencyAudit.estimateTransferSize(networkRecord, contentLength,
-        'Script');
-    const wastedRatio = (unusedLength / contentLength) || 0;
-    const wastedBytes = Math.round(totalBytes * wastedRatio);
+    const wastedRatio = (unused / content) || 0;
+    const wastedBytes = Math.round(lengths.transfer * wastedRatio);
 
     return {
-      url: networkRecord.url,
-      totalBytes,
+      url: url,
+      totalBytes: lengths.transfer,
       wastedBytes,
       wastedPercent: 100 * wastedRatio,
     };
   }
 
+  /**
+   * @param {WasteData[]} wasteData
+   * @param {LH.Artifacts.NetworkRequest} networkRecord
+   */
+  static determineLengths(wasteData, networkRecord) {
+    let unused = 0;
+    let content = 0;
+    // TODO: this is right for multiple script tags in an HTML document,
+    // but may be wrong for multiple frames using the same script resource.
+    for (const usage of wasteData) {
+      unused += usage.unusedLength;
+      content += usage.contentLength;
+    }
+    const transfer = ByteEfficiencyAudit.estimateTransferSize(networkRecord, content, 'Script');
+
+    return {
+      content,
+      unused,
+      transfer,
+    };
+  }
+
   /**
    * @param {LH.Artifacts} artifacts
    * @param {Array<LH.Artifacts.NetworkRequest>} networkRecords
-   * @return {ByteEfficiencyAudit.ByteEfficiencyProduct}
+   * @param {LH.Audit.Context} context
+   * @return {Promise<ByteEfficiencyAudit.ByteEfficiencyProduct>}
    */
-  static audit_(artifacts, networkRecords) {
+  static async audit_(artifacts, networkRecords, context) {
+    const bundles = artifacts.SourceMaps ? await JSBundles.request(artifacts, context) : [];
+    const {bundleSourceUnusedThreshold = 1024} = context.options || {};
```

Comment:
since we dont show bundles in the table if they're >=2048 bytes, it feels like we can set this threshold filter to be smaller... like 512... since we're just determining if we're gonna decorate the bundle with specific subrows, right?

Reply:
also can you move the const value dfn up to L22ish

Reply:
done